src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+        }
+    }
+    anchored_x = rect.x;
+    anchored_y = rect.y;
+}
+
+unsigned int ui_element::get_ax() const
+{
+    return anchored_x;
+}
+
+unsigned int ui_element::get_ay() const
+{
+    return anchored_y;
+}
+
+void ui_element::set_parent( const ui_window *parent )
+{
+    this->parent = parent;
+    calc_anchored_values();
+}
+
+const ui_rect &ui_element::get_rect() const
+{
+    return rect;
+}
+
+WINDOW *ui_element::get_win() const
+{
+    if( parent != nullptr ) {
+        return parent->get_win();
+    }
+    return nullptr;
+}
+
+ui_label::ui_label( std::string text ,int x, int y, ui_anchor anchor ) : ui_element( utf8_width( text.c_str() ), 1, x, y, anchor ), text( text )
+{
+}
+
+void ui_label::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), text_color, text.c_str() );
+}
+
+void ui_label::set_text( std::string new_text )
+{
+    text = new_text;
+    set_rect( ui_rect( new_text.size(), get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+bordered_window::bordered_window( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor ) : ui_window( size_x, size_y, x, y, anchor )
+{
+}
+
+void bordered_window::local_draw()
+{
+    auto win = get_win(); // never null
+
+    wattron( win, border_color );
+    wborder( win, LINE_XOXO, LINE_XOXO, LINE_OXOX, LINE_OXOX,
+            LINE_OXXO, LINE_OOXX, LINE_XXOO, LINE_XOOX );
+    wattroff( win, border_color );
+}
+
+health_bar::health_bar( size_t size_x, int x, int y, ui_anchor anchor ) : ui_element( size_x, 1, x, y, anchor ),
+                       max_health( size_x * points_per_char ), current_health( max_health )
+{
+    for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+        bar_str += "|";
+    }
+}
+
+void health_bar::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), bar_color, bar_str.c_str() );
+}
+
+void health_bar::refresh_bar( bool overloaded, float percentage )
+{
+    bar_str = "";
+    if( overloaded ) {
+        for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+            bar_str += "*";
+        }
+    } else {
+        for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+            int char_health = current_health - ( i * points_per_char );
+            if( char_health <= 0 ) {
+                bar_str += ".";
+            } else if( char_health == 1 ) {
+                bar_str += "\\";
+            } else {
+                bar_str += "|";
+            }
+        }
+    }
+
+    static const std::array<nc_color, 7> colors {{
+        c_green,
+        c_ltgreen,
+        c_yellow,
+        c_magenta,
+        c_ltred,
+        c_red,
+        c_ltgray
+    }};
+
+    bar_color = colors[(int) roundf( colors.size() * percentage )];
+}
+
+void health_bar::set_health_percentage( float percentage )
+{
+    bool overloaded = false;
+    if( percentage > 1 ) {
```

Comment:
This handles `percentage` values outside the usual range, but `refresh_bar` (which is called from here) does not handle them (the last statement with the `roundf` will probably cause problems as it may produce an index outside the valid range).